src/Cooked/MockChain/Balancing.hs

```diff
@@ -1,409 +1,326 @@
-          | otherwise -> do
-              -- Debug.Trace.traceM $ "New iteration: newfee = " <> show newFee
-              calcFee (n - 1) newFee skel collateralUtxos
+  return (txSkelBal, fee, adjustedCollateralIns, returnCollateralWallet)
 
--- | This funcion is essentially a copy of
--- https://github.com/input-output-hk/plutus-apps/blob/d4255f05477fd8477ee9673e850ebb9ebb8c9657/plutus-ledger/src/Ledger/Fee.hs#L19
-estimateTxSkelFee ::
-  Emulator.Params ->
-  Map Api.DatumHash Api.Datum ->
-  Map Api.TxOutRef Api.TxOut ->
-  Map Script.ValidatorHash (Script.Versioned Script.Validator) ->
-  TxSkel ->
-  Fee ->
-  Set Api.TxOutRef ->
-  Either MockChainError Fee
-estimateTxSkelFee params managedData managedTxOuts managedValidators skel fees collateralIns = do
-  txBodyContent <-
-    left MCEGenerationError $
-      generateBodyContent fees collateralIns params managedData managedTxOuts managedValidators skel
-  let nkeys = Cardano.estimateTransactionKeyWitnessCount txBodyContent
-  txBody <-
-    left (MCEGenerationError . TxBodyError "Error creating body when estimating fees") $
-      Cardano.createAndValidateTransactionBody Cardano.ShelleyBasedEraConway txBodyContent
-  case Cardano.evaluateTransactionFee Cardano.ShelleyBasedEraConway (Emulator.pEmulatorPParams params) txBody nkeys 0 of
-    Emulator.Coin fee -> pure $ Fee fee
+-- | This computes the minimum and maximum possible fee a transaction can cost
+-- based on the current protocol parameters
+getMinAndMaxFee :: (MonadBlockChainBalancing m) => m (Fee, Fee)
+getMinAndMaxFee = do
+  -- Default parameters in case they are not present. It is unclear when/if this
+  -- could actually happen though.
+  let defMaxTxExecutionUnits =
+        Cardano.ExecutionUnits {executionSteps = 10_000_000_000, executionMemory = 14_000_000}
+      defExecutionUnitPrices =
+        Cardano.ExecutionUnitPrices {priceExecutionSteps = 721 Rat.% 10_000_000, priceExecutionMemory = 577 Rat.% 10_000}
+  -- Parameters necessary to compute the maximum possible fee for a transaction
+  params <- Emulator.pProtocolParams <$> getParams
+  let maxTxSize = toInteger $ Cardano.protocolParamMaxTxSize params
+      Emulator.Coin txFeePerByte = Cardano.protocolParamTxFeePerByte params
+      Emulator.Coin txFeeFixed = Cardano.protocolParamTxFeeFixed params
+      Cardano.ExecutionUnitPrices priceESteps priceEMem = fromMaybe defExecutionUnitPrices $ Cardano.protocolParamPrices params
+      Cardano.ExecutionUnits (toInteger -> eSteps) (toInteger -> eMem) = fromMaybe defMaxTxExecutionUnits $ Cardano.protocolParamMaxTxExUnits params
+  -- Final fee accounts for the size of the transaction and the units consumed
+  -- by the execution of scripts from the transaction
+  let sizeFees = txFeeFixed + (maxTxSize * txFeePerByte)
+      eStepsFees = (eSteps * Rat.numerator priceESteps) `div` Rat.denominator priceESteps
+      eMemFees = (eMem * Rat.numerator priceEMem) `div` Rat.denominator priceEMem
```

Comment:
I imagine these are respectively fees related to the size of the transaction, fees related to time consumption when validating (cpu cycles running the scripts), and fees related to space consumption when validating (memory used when running the scripts), right?

Reply:
that's correct